### PR TITLE
Allowed array/struct values in `default` definition of `declare`

### LINKED ISSUE #2119: Add support for arrays and structs in DECLARE...DEFAULT (bigquery dialect)
Bigquery dialect currently does not support arrays and structs values in `DEFAULT` section of `DECLARE`
causing `PRS` error
```sql
DECLARE arr1 ARRAY<STRING> DEFAULT ['aaaa', 'bbb', 'cccc'];
DECLARE str1 STRUCT<f1 STRING, f2 STRING, f3 STRING> DEFAULT ('aaaa', 'bbb', 'cccc');
```
lining result
```
L:   1 | P:  27 |  PRS | Line 1, Position 27: Found unparsable section: " DEFAULT ['aaaa', 'bbb', 'cccc']"
L:   2 | P:  53 |  PRS | Line 2, Position 53: Found unparsable section: " DEFAULT ('aaaa', 'bbb', 'cccc')"
```
parsing result
```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        declare_segment:
[L:  1, P:  1]      |            keyword:                                          'DECLARE'
[L:  1, P:  8]      |            whitespace:                                       ' '
[L:  1, P:  9]      |            identifier:                                       'arr1'
[L:  1, P: 13]      |            whitespace:                                       ' '
[L:  1, P: 14]      |            data_type:
[L:  1, P: 14]      |                keyword:                                      'ARRAY'
[L:  1, P: 19]      |                start_angle_bracket:                          '<'
[L:  1, P: 20]      |                [META] indent:
[L:  1, P: 20]      |                data_type:
[L:  1, P: 20]      |                    data_type_identifier:                     'STRING'
[L:  1, P: 26]      |                [META] dedent:
[L:  1, P: 26]      |                end_angle_bracket:                            '>'
[L:  1, P: 27]      |            unparsable:                                       !! Expected: 'Nothing...'
[L:  1, P: 27]      |                whitespace:                                   ' '
[L:  1, P: 28]      |                raw:                                          'DEFAULT'
[L:  1, P: 35]      |                whitespace:                                   ' '
[L:  1, P: 36]      |                start_square_bracket:                         '['
[L:  1, P: 37]      |                raw:                                          "'aaaa'"
[L:  1, P: 43]      |                comma:                                        ','
[L:  1, P: 44]      |                whitespace:                                   ' '
[L:  1, P: 45]      |                raw:                                          "'bbb'"
[L:  1, P: 50]      |                comma:                                        ','
[L:  1, P: 51]      |                whitespace:                                   ' '
[L:  1, P: 52]      |                raw:                                          "'cccc'"
[L:  1, P: 58]      |                end_square_bracket:                           ']'
[L:  1, P: 59]      |    statement_terminator:                                     ';'
[L:  1, P: 60]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    statement:
[L:  2, P:  1]      |        declare_segment:
[L:  2, P:  1]      |            keyword:                                          'DECLARE'
[L:  2, P:  8]      |            whitespace:                                       ' '
[L:  2, P:  9]      |            identifier:                                       'str1'
[L:  2, P: 13]      |            whitespace:                                       ' '
[L:  2, P: 14]      |            data_type:
[L:  2, P: 14]      |                keyword:                                      'STRUCT'
[L:  2, P: 20]      |                start_angle_bracket:                          '<'
[L:  2, P: 21]      |                [META] indent:
[L:  2, P: 21]      |                parameter:                                    'f1'
[L:  2, P: 23]      |                whitespace:                                   ' '
[L:  2, P: 24]      |                data_type:
[L:  2, P: 24]      |                    data_type_identifier:                     'STRING'
[L:  2, P: 30]      |                comma:                                        ','
[L:  2, P: 31]      |                whitespace:                                   ' '
[L:  2, P: 32]      |                parameter:                                    'f2'
[L:  2, P: 34]      |                whitespace:                                   ' '
[L:  2, P: 35]      |                data_type:
[L:  2, P: 35]      |                    data_type_identifier:                     'STRING'
[L:  2, P: 41]      |                comma:                                        ','
[L:  2, P: 42]      |                whitespace:                                   ' '
[L:  2, P: 43]      |                parameter:                                    'f3'
[L:  2, P: 45]      |                whitespace:                                   ' '
[L:  2, P: 46]      |                data_type:
[L:  2, P: 46]      |                    data_type_identifier:                     'STRING'
[L:  2, P: 52]      |                [META] dedent:
[L:  2, P: 52]      |                end_angle_bracket:                            '>'
[L:  2, P: 53]      |            unparsable:                                       !! Expected: 'Nothing...'
[L:  2, P: 53]      |                whitespace:                                   ' '
[L:  2, P: 54]      |                raw:                                          'DEFAULT'
[L:  2, P: 61]      |                whitespace:                                   ' '
[L:  2, P: 62]      |                bracketed:
[L:  2, P: 62]      |                    start_bracket:                            '('
[L:  2, P: 63]      |                    raw:                                      "'aaaa'"
[L:  2, P: 69]      |                    comma:                                    ','
[L:  2, P: 70]      |                    whitespace:                               ' '
[L:  2, P: 71]      |                    raw:                                      "'bbb'"
[L:  2, P: 76]      |                    comma:                                    ','
[L:  2, P: 77]      |                    whitespace:                               ' '
[L:  2, P: 78]      |                    raw:                                      "'cccc'"
[L:  2, P: 84]      |                    end_bracket:                              ')'
[L:  2, P: 85]      |    statement_terminator:                                     ';'
[L:  2, P: 86]      |    newline:                                                  '\n'
```

### PR DESCRIPTION
**Pull request by KulykDmytro**
closes #2119 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by KulykDmytro** [comment]
seems like `Tuple` segment created in PR needs to be moved into common/ansi section

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_bigquery.py:600]: Think this match grammar should just be
```sql
match_grammar = Bracketed(
        Delimited(
            Sequence(
                Ref("BaseExpressionElementGrammar"),
                Ref("AliasExpressionSegment", optional=True),
            ),
        ),
    )
```
that's what's in the docs
image

**Review by jpy-git — COMMENTED** [review]
@KulykDmytro the tuple segment can be simplified, but otherwise it looks good!

**Comment by jpy-git** [comment]
> seems like `Tuple` segment created in PR needs to be moved into common/ansi section

Which other dialects will we need to use it in?

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (KulykDmytro) [reply src/sqlfluff/dialects/dialect_bigquery.py:600]: tuple can be named
```
('aaaa' as f1, 'bb' as f2)
```

Comment (KulykDmytro) [reply src/sqlfluff/dialects/dialect_bigquery.py:600]: agree, will change

Comment (KulykDmytro) [reply src/sqlfluff/dialects/dialect_bigquery.py:600]: will revert `TypelessStruct`
and redefine `Tuple`

**Comment by KulykDmytro** [comment]
> > seems like `Tuple` segment created in PR needs to be moved into common/ansi section
> 
> Which other dialects will we need to use it in?

i beleive this one is basic `ansi` syntax

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_bigquery.py:611]: what was wrong with just changing the match grammar in tuple to what I had in my original comment?

Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_bigquery.py:611]: Changing TypelessStruct was fine as you originally did it, its just the original match grammar was too verbose.

Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_bigquery.py:611]: > what about any number of values in tuple?

Delimited contains contains any number of items anyway so the AnyNumber of was redundant

Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_bigquery.py:611]: when I tried it, it parsed fine for all the existing test cases

Comment (KulykDmytro) [reply src/sqlfluff/dialects/dialect_bigquery.py:611]: TypelessStruct and Tuple syntax are partially compatible as of BQ documentation

**Comment by KulykDmytro** [comment]
still cant cover a case when type is set after `default` keyword
```
declare str1 default struct<f1 string, f2 string>('one', 'two');
```
based on original syntax `data_type` is optional
image

**Comment by jpy-git** [comment]
If that should pass then add it to the unit tests, along with any other variations that should parse. Otherwise it's hard to review

**Comment by KulykDmytro** [comment]
is it ok to do like this? to avoid duplication
```
    _default_section = Sequence(
        "DEFAULT",
        OneOf(
            Ref("LiteralGrammar"),
            Bracketed(Ref("SelectStatementSegment")),
            Ref("BareFunctionSegment"),
            Ref("FunctionSegment"),
            Ref("ArrayLiteralSegment"),
            Ref("TypelessStructSegment"),
            Ref("TupleSegment"),
        ),
        optional=True,
    )
    parse_grammar = Sequence(
        "DECLARE",
        Delimited(Ref("NakedIdentifierSegment")),
        OneOf(
            Sequence(
                Ref("DatatypeSegment"),
                _default_section,
            ),
            _default_section,
        ),
    )
```

**Comment by jpy-git** [comment]
@KulykDmytro 

That's just
```sql
    parse_grammar = Sequence(
        "DECLARE",
        Delimited(Ref("NakedIdentifierSegment")),
        Ref("DatatypeSegment", optional=True),
        Sequence(
            "DEFAULT",
            OneOf(
                Ref("LiteralGrammar"),
                Bracketed(Ref("SelectStatementSegment")),
                Ref("BareFunctionSegment"),
                Ref("FunctionSegment"),
                Ref("ArrayLiteralSegment"),
                Ref("TypelessStructSegment"),
                Ref("TupleSegment"),
            ),
            optional=True,
        )
    )
```

**Comment by KulykDmytro** [comment]
not really. in this case we'll have two optional blocks which causing following:
`default` keyword being interpreted as `data_type`:
```
declare var1 default 'value';
----
- statement:
    declare_segment:
      keyword: declare
      identifier: var1
      data_type:
        data_type_identifier: default  <<<<
      unparsable:
        raw: "'value'"
- statement_terminator: ;
```

**Comment by KulykDmytro** [comment]
@jpy-git can you please assist with mypy error
Can't even imagine what goes wrong
```
mypy run-test: commands[0] | mypy src/sqlfluff
src/sqlfluff/rules/L003.py:347: error: Unsupported operand types for + ("Tuple[RawSegment, ...]" and "Tuple[BaseSegment]")

```

**Comment by jpy-git** [comment]
@KulykDmytro don't worry it's nothing to do with your work, mypy had a new release this morning and the new version is flagging errors in code already in main. All the PRs are failing at the moment due to the error. I've fixed it in https://github.com/sqlfluff/sqlfluff/pull/2125 so just waiting for review and merge and then you can update your branch from main.

**Comment by jpy-git** [comment]
> not really. in this case we'll have two optional blocks which causing following:
> `default` keyword being interpreted as `data_type`

Makes sense, that should parse but it's a known issue with the fact that our datatype identifier segment can capture keywords atm since most dialects haven't fully implemented all their datatypes (discussed [here](https://github.com/sqlfluff/sqlfluff/pull/2121#discussion_r770772258)

Your approach should be fine and the tests look pretty thorough 👍

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_bigquery.py:814]: just removing a redundant OneOf so fine with this 👍

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (jpy-git) [inline src/sqlfluff/dialects/dialect_bigquery.py:792]: @tunetheweb @barrywhart question here around storing some reused grammar in the `_default_section` attribute. The logic is fine, just wondering if you prefer this split into a grammar/segment rather than an attribute for consistency with the rest of the codebase.

A grammar in `bigquery_dialect.add(` (line 79) sounds sensible to me.

**Review by jpy-git — COMMENTED** [review]
@KulykDmytro just checking with the others on where they want that that reused grammar, but otherwise looks good and happy to approve once that's resolved 👍

File: src/sqlfluff/dialects/dialect_bigquery.py
Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_bigquery.py:792]: I prefer a grammar as well. As you say, it's mainly about consistency to avoid confusing future contributors.

Comment (jpy-git) [reply src/sqlfluff/dialects/dialect_bigquery.py:792]: makes sense!

@KulykDmytro if you take `_default_section` and add it to here: https://github.com/sqlfluff/sqlfluff/blob/bf12e8b2371fdcfc617282256c2d9d2d4d9e9f7e/src/sqlfluff/dialects/dialect_bigquery.py#L79
like:
```
DefaultDeclareOptionsGrammar = Sequence(
        "DEFAULT",
        OneOf(
            Ref("LiteralGrammar"),
            Bracketed(Ref("SelectStatementSegment")),
            Ref("BareFunctionSegment"),
            Ref("FunctionSegment"),
            Ref("ArrayLiteralSegment"),
            Ref("TypelessStructSegment"),
            Ref("TupleSegment"),
            Ref("BaseExpressionElementGrammar"),
        ),
    )
```
and then you can just reference that grammar in your code using `Ref("DefaultDeclareOptionsGrammar")`

**Review by jpy-git — APPROVED** [review]
Looks good, nice work 👍

**Pull request merged**
